### PR TITLE
[Maint] Add zarr to optional dependencies in pyproject.toml

### LINKED ISSUE #2321: Missing dependencies for drag-and-dropping zarr array
## 🐛 Bug

Missing dependencies for drag-and-dropping a persistent `zarr` array from disk.

## To Reproduce
Drag-and-dropping a persistent `zarr` array into a fresh install of `napari` leads to the following errors:

```
16:39:23 ERROR PluginError: Error in plugin 'builtins', hook 'napari_get_reader'
  Cause was: ModuleNotFoundError("No module named 'zarr'")
    in file: /usr/local/Caskroom/miniconda/base/envs/napari_zarr_bug/lib/python3.8/site-packages/dask/array/core.py
    at line: 3161
     author: napari team
      email: napari-steering-council@googlegroups.com
    package: napari
        url: https://napari.org
    version: 0.4.5
```

After installing `zarr`, there is another missing dependency:
```
16:39:48 ERROR PluginError: Error in plugin 'builtins', hook 'napari_get_reader'
  Cause was: ImportError("fsspec is required to use any file-system functionality. Please install using\nconda install -c conda-forge 'fsspec>=0.3.3'\nor\npython -m pip install 'fsspec>=0.3.3'")
    in file: /usr/local/Caskroom/miniconda/base/envs/napari_zarr_bug/lib/python3.8/site-packages/dask/bytes/__init__.py
    at line: 9
     author: napari team
      email: napari-steering-council@googlegroups.com
    package: napari
        url: https://napari.org
    version: 0.4.5
```

Running the suggested `pip install 'fsspec>=0.3.3'` solves the issue.

## Expected behavior

It seems reasonable that a user who whishes to open `zarr` arrays with `napari` has to install `zarr` first.
However, it's not obvious that one has to install `fsspec` for this, therefore `fsspec` should be a dependency of `napari`.

## Environment

napari: 0.4.5
Platform: macOS-10.15.7-x86_64-i386-64bit
System: MacOS 10.15.7
Python: 3.8.8 (default, Feb 24 2021, 13:46:16) [Clang 10.0.0 ]
Qt: 5.15.2
PyQt5: 5.15.3
NumPy: 1.20.1
SciPy: 1.6.1
Dask: 2021.02.0
VisPy: 0.6.6

OpenGL:
- GL version: 2.1 ATI-3.10.19
- MAX_TEXTURE_SIZE: 16384

Screens:
- screen 1: resolution 1792x1120, scale 2.0
- screen 2: resolution 2560x1440, scale 2.0

Plugins:
- console: 0.0.3
- svg: 0.1.4

### PR DESCRIPTION
**Pull request by psobolewskiPhD**
# References and relevant issues
Closes https://github.com/napari/napari/issues/2321

# Description

`builtins` uses `dask.array.from_zarr` to open zarrs.
napari depends on `dask[array]`, but `zarr` is an *optional* dependency that is not installed (see https://docs.dask.org/en/stable/install.html#optional-dependencies)

This PR adds the dask optional zarr requirement to napari requirements, such that zarr files work with builtins.
Note: this is *already the case* for our conda-forge package:
https://github.com/conda-forge/napari-feedstock/blob/32a4eb04ca7b6ccd2c4e146bde204f1dd5425a17/recipe/meta.yaml#L81

Also, this will likely be needed anyway if we want to use zarr for dealing with large label arrays, see:
https://github.com/napari/napari/issues/470

### CODE REVIEW

**Comment by Czaki** [comment]
Att his moment I prefer to find way to remove dask from default dependencies. I do not like idea of adding zarr. 

It could be in all, but I do not like idea of it in basic dependencies.

**Comment by psobolewskiPhD** [comment]
Feels like napari should open a zarr out of the box, but for this builtins doesn't need to use use dask.from_zarr 
Could just use zarr.open ?

**Comment by brisvag** [comment]
@napari-bot update constraints

**Comment by brisvag** [comment]
> @napari-bot update constraints

yay, fixed the constraint updater!

File: setup.cfg [hunk 1]
```diff
@@ -79,6 +79,7 @@ install_requires =
     typing_extensions>=4.2.0
     vispy>=0.13.0,<0.14
     wrapt>=1.11.1
+    zarr>=2.12.0 # needed by `builtins` (dask.array.from_zarr) to open zarr
```
Comment (Carreau) [inline setup.cfg]: is there a way to depends on `dask[array]`, so that say if they update their dependencies it reflects here ?

Comment (psobolewskiPhD) [reply setup.cfg]: I don't think so?
`dask[array]` doesn't depend on zarr and `pip install "dask[array]"` in a fresh python 3.10:
```
Installing collected packages: zipp, toolz, pyyaml, packaging, numpy, locket, fsspec, cloudpickle, click, partd, importlib-metadata, dask
```

Comment (Carreau) [reply setup.cfg]: Oh, ok I misunderstood.I see that `from_zarr` does a lazy import of `zarr` indeed.

I think as a desktop app for end users then it's reasonable to install zarr by default.

Comment (Czaki) [reply setup.cfg]: Yes. for bundled version I will add zarr dependency

**Comment by psobolewskiPhD** [comment]
Bumping this, fell through the cracks of 0.4.19, but we have 2nd issue raising this:
https://github.com/napari/napari/issues/6708
And plans to have smarter labels layer:
https://github.com/napari/napari/pull/6190/files

File: pyproject.toml [hunk 2]
```diff
@@ -72,6 +72,7 @@ dependencies = [
     "typing_extensions>=4.2.0",
     "vispy>=0.14.1,<0.15",
     "wrapt>=1.11.1",
+    "zarr zarr>=2.12.0", # needed by `builtins` (dask.array.from_zarr) to open zarr
```
Comment (psobolewskiPhD) [inline pyproject.toml]:
```suggestion
    "zarr>=2.12.0", # needed by `builtins` (dask.array.from_zarr) to open zarr
```

**Comment by Czaki** [comment]
you need to update title of PR

**Comment by jni** [comment]
Oof. This is tricky. I basically agree with @Czaki saying it would be better for these things to go in `[all]`. To make the counterpoint to @Carreau's

> I think as a desktop app for end users then it's reasonable to install zarr by default.

napari is both a desktop app *and* a library, and I think for the library case we have been a little too lackadaisical in taking on dependencies. Yes there's issues where some functionality doesn't work if you don't install the dependency, but they can be easily fixed by installing the dependency. (And of course in the bundled app we would use `[all]`.) The converse problems — dependency conflicts, slow import/startup, bloated environments — are not so easily resolved, and we also have plenty of reports of that (including in the survey). So I'm definitely leaning towards zarr being optional too.

**Comment by DragaDoncila** [comment]
> napari is both a desktop app and a library, and I think for the library case we have been a little too lackadaisical in taking on dependencies

I think if `builtins` doesn't work without `zarr`, and `builtins` is available for use by the library version of `napari`, then we need to depend on `zarr`...

**Comment by jni** [comment]
> If we don't then we should consider removing any kind of plugin usage from the library but that's a little wild.

is it though? I actually don't think it's that crazy to put builtins in [all]. Nobody expects that you should be able to drag and drop a csv into a matplotlib figure and get a plot, and I think napari *should* aim to be a matplotlib-adjacent library.

It's a shift in mindset, sure, but it's one that is perfectly in line with napari-lite etc.

And anyway I don't think builtins is exactly tied to zarr, it could go back to just being for tiffs/imageio, which are much more mature libraries. (But even there, I think they are excessive for the base library.)

We could of course invert things on the package-name front, where our default package is `napari-base` and `napari` pulls in `napari-base[all]`. But that is an accounting trick that doesn't change the conversation here — what is the minimal set of dependencies for napari to actually function as a viewer in Python. I don't think zarr is one.

**Comment by psobolewskiPhD** [comment]
The thing is napari has advertised drag-and-drop to open stuff.
And the homepage has:
image

Hence we get issues of `why doesn't zarr work`. Heck *I* find this annoying as I make envs a lot and always have to remember to install zarr.

At this point, if we want to drop zarr support I think it should be with clear messaging to users and a deprecation period.

**Comment by DragaDoncila** [comment]
Hmm yeah so it should **definitely** be in `[all]`, and my reasoning is that it should be available to `viewer.open` based on our current ethos - but as @jni points out, as we move to `napari-lite`, maybe we actually strip plugins from the library, or make them a separate extra. And maybe we don't want an in the middle bit where we just add `zarr` and then drop it, so I guess i'm in the `[all]` basket.

**Comment by Czaki** [comment]
> I think if `builtins` doesn't work without `zarr`, and `builtins` is available for use by the library version of `napari`, then we need to depend on `zarr`... so for that reason I don't think `all` makes sense, because `all` is for people who want the `app`, but we want file opening to be present in the `library`, no? If we don't then we should consider removing **any** kind of plugin usage from the library but that's a little wild.

Pandas requires openpyxl to work with Excel files, pyarrow for other utils. It is nothing unexpected that some library utils requires some additional dependencies. 

But I vote to put this in `optional`, not `all`. After merge #6720 all things listed in `optional` will be installed from `all` but `optional` does not enforce qt backend.

**Comment by psobolewskiPhD** [comment]
I'm not a fan of looking at everything through the prism of a very much hypothetical napari-lite.
Feels like a lot of overcomplicating, I mean if `builtins` was a plugin, it would depend on zarr, so...

**Comment by jni** [comment]
> but optional does not enforce qt backend.

Oh I thought we'd removed pyqt from all, oops. 😂 I think I was confused because of removing pyqt from the conda recipe. 😅

**Comment by psobolewskiPhD** [comment]
Closing, see rationale here:
https://github.com/napari/napari/pull/6190#issuecomment-1986652594

**Comment by jni** [comment]
@psobolewskiPhD I didn't want you to close it, I wanted you to move it to optional! 😂

**Comment by jni** [comment]
(Part of my argument is that there are middle roads that we *should* take!)

**Review by jni — APPROVED** [review]

**Review by brisvag — APPROVED** [review]

**Review by Czaki — APPROVED** [review]

**Pull request merged**